File: client/src/main/scala/com/alphasystem/game/uno/client/ui/control/CardView.scala
package com.alphasystem.game.uno.client.ui.control

import com.alphasystem.game.uno.client.ui.control.delegate.{CardView => JCardView}
import com.alphasystem.game.uno.model.Card
import javafx.beans.property
import scalafx.Includes._
import scalafx.beans.property._
import scalafx.scene.Node

class CardView private(override val delegate: JCardView) extends Node(delegate) {

  def this(card: Card, playerName: String, height: Int) = this(new JCardView(card, playerName, height))

  def selected: BooleanProperty = delegate.selectedProperty

  def selected_=(selected: Boolean): Unit = delegate.setSelected(selected)

  def card: ObjectProperty[Card] = delegate.cardProperty()

  def card_=(card: Card): Unit = delegate.setCard(card)

  def playerName: property.StringProperty = delegate.playerNameProperty()

  def playerName_=(playerName: String): Unit = delegate.setPlayerName(playerName)

  def fitHeight: Int = delegate.getFitHeight

  def fitHeight_=(height: Int): Unit = delegate.setFitHeight(height)
}

object CardView {
  def apply(): CardView = CardView(null, null)

  def apply(card: Card, playerName: String, height: Int = 128): CardView = new CardView(card, playerName, height)
}

File: client/src/main/java/com/alphasystem/game/uno/client/ui/control/delegate/CardView.java
package com.alphasystem.game.uno.client.ui.control.delegate;

import com.alphasystem.game.uno.model.Card;
import javafx.beans.property.*;
import javafx.scene.control.Control;
import javafx.scene.control.Skin;

public final class CardView extends Control {

    private final BooleanProperty selected = new SimpleBooleanProperty(this, "selected", false);
    private final ObjectProperty<Card> card = new SimpleObjectProperty<>(this, "card", null);
    private final StringProperty playerName = new SimpleStringProperty(this, "playerName", null);
    private final IntegerProperty fitHeight = new SimpleIntegerProperty(this, "fitHeight", 128);

    public CardView() {
        this(null, null, 128);
    }

    public CardView(Card card) {
        this(card, null, 128);
    }

    public CardView(Card card, String playerName, int height) {
        setSkin(createDefaultSkin());
        setCard(card);
        setPlayerName(playerName);
        setSelected(false);
        setFitHeight(height);
    }

    @Override
    protected Skin<?> createDefaultSkin() {
        return new CardSkin(this);
    }

    public boolean isSelected() {
        return selected.get();
    }

    public BooleanProperty selectedProperty() {
        return selected;
    }

    public void setSelected(boolean selected) {
        this.selected.set(selected);
    }

    public Card getCard() {
        return card.get();
    }

    public ObjectProperty<Card> cardProperty() {
        return card;
    }

    public void setCard(Card card) {
        this.card.set(card);
    }

    public String getPlayerName() {
        return playerName.get();
    }

    public StringProperty playerNameProperty() {
        return playerName;
    }

    public void setPlayerName(String playerName) {
        this.playerName.set(playerName);
    }

    public int getFitHeight() {
        return fitHeight.get();
    }

    public IntegerProperty fitHeightProperty() {
        return fitHeight;
    }

    public void setFitHeight(int fitHeight) {
        this.fitHeight.set(fitHeight);
    }
}
